src/profile-logic/process-profile.js

```diff
@@ -223,26 +252,114 @@ function processThread(thread, libs) {
             );
           }
           fileName = stringTable.indexForString(scriptURI);
-          lineNumber = jsMatch[3] | 0;
+          lineNumber = parseInt(jsMatch[3], 10);
         }
       }
     }
     funcIndex = funcTable.length;
-    addFunc(
-      funcNameIndex,
-      resourceIndex,
-      addressRelativeToLib,
-      isJS,
-      fileName,
-      lineNumber
-    );
+    {
+      // Add the function to the funcTable.
+      const index = funcTable.length++;
+      funcTable.name[index] = funcNameIndex;
+      funcTable.resource[index] = resourceIndex;
+      funcTable.address[index] = addressRelativeToLib;
+      funcTable.isJS[index] = isJS;
+      funcTable.fileName[index] = fileName;
+      funcTable.lineNumber[index] = lineNumber;
+    }
     stringTableIndexToNewFuncIndex.set(locationIndex, funcIndex);
     return funcIndex;
   });
-  frameTable.address = frameTable.func.map(
-    funcIndex => funcTable.address[funcIndex]
+
+  return [funcTable, resourceTable, frameFuncs];
+}
+
+/**
+ * Explicitly recreate the frame table here to help enforce our assumptions about types.
+ */
+function _processFrameTable(
+  geckoFrameTable: GeckoFrameStruct,
+  funcTable: FuncTable,
+  frameFuncs: IndexIntoFuncTable[]
+): FrameTable {
+  return {
+    address: frameFuncs.map(funcIndex => funcTable.address[funcIndex]),
+    category: geckoFrameTable.category,
+    func: frameFuncs,
+    implementation: geckoFrameTable.implementation,
+    line: geckoFrameTable.line,
+    optimizations: geckoFrameTable.optimizations,
+    length: geckoFrameTable.length,
+  };
+}
+
+/**
+ * Explicitly recreate the stack table here to help enforce our assumptions about types.
+ */
+function _processStackTable(geckoStackTable: GeckoStackStruct): StackTable {
```

Comment:
I was trying to remove most of the logic in `_processThread` and have other functions do the work of converting the tables. With this approach there is a clean break of where the work is done, even if some of it is currently trivial work. I do have a follow-up patch modifying `_processStackTable` and it is nice to be able to do that work in an encapsulated function.